dj/construction/build.py

```diff
@@ -492,52 +502,98 @@ def build_metric_nodes(
+                    f"`{col}`, which is not available on every"
+                    " metric and thus cannot be included.",
+                )
+
+    combined_ast: ast.Query = ast.Query(
+        select=ast.Select(from_=ast.From(relations=[])),
+        ctes=[],
     )
-    metric_dependencies: Set[str] = set()
-    for metric_node in metric_nodes[1:]:
-        if metric_node.type != NodeType.METRIC:  # pragma: no cover
-            raise DJInvalidInputException(
-                "Cannot build a query for multiple nodes if one or more "
-                f"of them aren't metric nodes. ({metric_node.name} is not"
-                "a metric node)",
-            )
+    initial_dimension_columns = []
+    all_dimension_columns = []
+    for (idx, metric_node) in enumerate(metric_nodes):
+        # Build each metric node separately
         metric_ast = build_node(
-            session,
-            metric_node.current,
-            filters,
-            dimensions,
+            session=session,
+            node=metric_node.current,
+            filters=filters,
+            dimensions=dimensions,
             build_criteria=build_criteria,
         )
-        metric_dependencies = metric_dependencies.union(
-            {tbl.alias_or_name.name for tbl in metric_ast.find_all(ast.Table)},
-        )
-        built_source_tables = {
-            tbl.alias_or_name.name for tbl in metric_ast.find_all(ast.Table)
-        }
-        diff_columns = set(combined_ast.select.projection).difference(
-            metric_ast.select.projection,
-        )
-        if all(tbl in built_source_tables for tbl in metric_dependencies):
-            metric_ast.select.projection.extend(diff_columns)
-            combined_ast = metric_ast
+
+        # Add the WITH statements to the combined query
+        metric_ast_alias = ast.Name(f"m{idx}_" + metric_node.name.replace(".", "_DOT_"))
+        metric_ast.alias = metric_ast_alias
+        metric_ast.parenthesized = True
+        metric_ast.as_ = True
+        combined_ast.ctes += [metric_ast]
+
+        # Add the metric and dimensions to the final query layer's SELECT
+        current_table = ast.Table(metric_ast_alias)
+        final_select_columns = [
+            ast.Column(
+                name=col.alias_or_name,  # type: ignore
+                _table=current_table,
+                _type=col.type,  # type: ignore
+            )
+            for col in metric_ast.select.projection
+        ]
+        metric_columns = [
+            col
+            for col in final_select_columns
+            if str(col.alias_or_name) not in dimensions
+        ]
+        dimension_columns = [
+            col for col in final_select_columns if str(col.alias_or_name) in dimensions
+        ]
+        all_dimension_columns += dimension_columns
+        combined_ast.select.projection.extend(metric_columns)
+
+        if not combined_ast.select.from_.relations:  # type: ignore
```

Comment:
What does this condition mean?

Reply:
Ah so these two conditions are trying to build the final select that joins together all the CTEs. This condition is checking if anything is in the select statement's `FROM` clause, and in the very first iteration through the CTEs, there won't be any, so we want to add that as the `FROM ...` relation. Then, the subsequent CTEs need to be joined in via `JOIN <first relation> ON ...`, which is where the `else` comes in.